design/storage-class-device-set.md

```diff
@@ -0,0 +1,438 @@
+                operator: In
+                values:
+                - us-west-1c
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 100
+            podAffinityTerm:
+              labelSelector:
+                matchExpressions:
+                - key: "rook.io/cluster"
+                  operator: In
+                  values:
+                    - cluster1
+      volumeClaimTemplates:
+      - resources:
+          requests:
+            storage: 5Ti
+        storageClassName: gp2-ebs
+```
+
+In this example, we've added a `nodeAffinity` to the `placement` that restricts
+all OSD Pods to a specific AZ. This case is only really useful if you specify
+multiple `StorageClassDeviceSets` for different AZs, so that has been done here.
+
+#### Example 3: Different resource needs
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    deviceSets:
+        - name: cluster1-set1
+          count: 3
+          resources:
+            requests:
+              cpu: 2
+              memory: 4Gi
+          placement:
+            podAntiAffinity:
+              preferredDuringSchedulingIgnoredDuringExecution:
+              - weight: 100
+                podAffinityTerm:
+                  labelSelector:
+                    matchExpressions:
+                    - key: "rook.io/cluster"
+                      operator: In
+                      values:
+                        - cluster1
+                    topologyKey: "failure-domain.beta.kubernetes.io/zone"
+          volumeClaimTemplates:
+          - resources:
+              requests:
+                storage: 5Ti
+            storageClassName: gp2-ebs
+    - name: cluster1-set2
+      count: 3
+      resources:
+        requests:
+          cpu: 2
+          memory: 8Gi
+      placement:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 100
+            podAffinityTerm:
+              labelSelector:
+                matchExpressions:
+                - key: "rook.io/cluster"
+                  operator: In
+                  values:
+                    - cluster1
+                topologyKey: "failure-domain.beta.kubernetes.io/zone"
+      volumeClaimTemplates:
+      - resources:
+          requests:
+            storage: 10Ti
+        storageClassName: gp2-ebs
+```
+
+In this example, we have two `StorageClassDeviceSets` with different capacities
+for the devices in each set. The devices with larger capacities would require a
+greater amount of memory to operate the OSD Pods, so that is reflected in the
+`resources` field.
+
+#### Example 4: Simple local storage
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    nodes:
+    - name: node1
+    - name: node2
+    - name: node3
+    deviceSets:
+    - name: cluster1-set1
+      count: 3
+      volumeClaimTemplates:
```

Comment:
None of these examples use more than a single volume claim template, and I'm struggling to imagine how multiple templates might work under the hood. Could you explain how multiple templates would be handled?

Reply:
In the revamped design here, you would have one Deployment and one Pod per OSD and you just configure the Pod to attach to two PVCs. Then you would have to specify some additional configuration on how you want the devices to be consumed (e.g. one for metadata and one for data).